### PR TITLE
feat: hdbscan tuning

### LINKED ISSUE #744: [UI] move cluster re-calculation into the point cloud store

### LINKED ISSUE #745: [UI] assign colors to clusters while the parameters are being adjusted

### PR DESCRIPTION
**Pull request by mikeldking**
resolves #745 
resolves #744 

Adds HDBSCAN tuning to the point-cloud. 

NB: there are still some optimizations that need to be made for the point-cloud to not reset the zoom.

https://github.com/Arize-ai/phoenix/assets/5640648/70d20e7f-6a0b-45f6-b7a6-efe538803f9e

### CODE REVIEW

File: app/package.json [hunk 1]
```diff
@@ -33,6 +33,7 @@
     "shade": "^0.1.1",
     "three": "0.139",
     "three-stdlib": "^2.17.3",
+    "use-deep-compare-effect": "^1.8.1",
```
Comment (mikeldking) [inline app/package.json:36]: Object equality is becoming harder to check. Adding this to make future dev easier.

File: app/src/components/pointcloud/PointCloudClusters.tsx [hunk 2]
```diff
@@ -29,55 +23,116 @@ type PointCloudClustersProps = {
...
+  );
   // const { selectedClusterId } = usePointCloud();
   // Keep a map of event id to position for fast lookup
   const eventPositionMap = useMemo(() => {
     return points.reduce((acc, point) => {
-      acc[point.metaData.eventId] = point.position;
+      acc[point.eventId] = point.position;
       return acc;
-    }, {} as Record<string, ThreeDimensionalPointItem["position"]>);
+    }, {} as Record<string, ThreeDimensionalPointItem["position"] | undefined>);
```
Comment (mikeldking) [inline app/src/components/pointcloud/PointCloudClusters.tsx:42]: force undefined checks

File: app/src/components/pointcloud/PointCloudClusters.tsx [hunk 3]
```diff
@@ -29,55 +23,116 @@ type PointCloudClustersProps = {
...
+          }))
+          .filter(
+            (d): d is { position: ThreeDimensionalPoint } => d.position != null
+          );
+        return {
+          ...cluster,
+          data: positionData,
+        };
+      })
+      .filter((cluster) => cluster.data.length > 0); // Remove empty clusters
```
Comment (mikeldking) [inline app/src/components/pointcloud/PointCloudClusters.tsx:62]: React reconciliation is staggered. Need to remove empty clusters.

File: app/src/components/pointcloud/ClusteringSettings.tsx [hunk 4]
```diff
@@ -185,6 +190,10 @@ export default function ClusteringSettings() {
           control={control}
           rules={{
             required: "field is required",
+            min: {
+              value: 0,
+              message: "must be a positive number",
```
Comment (RogerHYang) [inline app/src/components/pointcloud/ClusteringSettings.tsx:195]:
```suggestion
              message: "must be a non-negative number",
```

**Review by RogerHYang — APPROVED** [review]

**Pull request merged**